trustpinning/certs_test.go

```diff
@@ -203,11 +208,274 @@ func TestValidateRoot(t *testing.T) {
+	certStore.RemoveAll()
+	goodRootCABundle, err := trustmanager.LoadCertBundleFromFile(validCAFilepath)
+	require.NoError(t, err)
+	memKeyStore := trustmanager.NewKeyMemoryStore(passphraseRetriever)
+	cryptoService := cryptoservice.NewCryptoService(memKeyStore)
+	testPubKey, err := cryptoService.Create("root", "notary-signer", data.ECDSAKey)
+	require.NoError(t, err)
+	testPrivKey, _, err := memKeyStore.GetKey(testPubKey.ID())
+	require.NoError(t, err)
+	expiredCert, err := generateExpiredTestingCertificate(testPrivKey, "notary-signer")
+	require.NoError(t, err)
+	bundleWithExpiredCert, err := trustmanager.CertChainToPEM(append(goodRootCABundle, expiredCert))
+	require.NoError(t, err)
+	bundleWithExpiredCertPath := filepath.Join(tempBaseDir, "bundle_with_expired_cert.pem")
+	require.NoError(t, ioutil.WriteFile(bundleWithExpiredCertPath, bundleWithExpiredCert, 0644))
+
+	// Check that we validate correctly against a pinned CA and provided bundle
+	err = ValidateRoot(certStore, newTestSignedRoot, "notary-signer", TrustPinConfig{CA: map[string]string{"notary-signer": bundleWithExpiredCertPath}, DisableTOFU: true})
+	require.NoError(t, err)
+
+	certStore.RemoveAll()
+	testPubKey2, err := cryptoService.Create("root", "notary-signer", data.ECDSAKey)
+	require.NoError(t, err)
+	testPrivKey2, _, err := memKeyStore.GetKey(testPubKey2.ID())
+	require.NoError(t, err)
+	expiredCert2, err := generateExpiredTestingCertificate(testPrivKey2, "notary-signer")
+	require.NoError(t, err)
+	allExpiredCertBundle, err := trustmanager.CertChainToPEM([]*x509.Certificate{expiredCert, expiredCert2})
+	require.NoError(t, err)
+	allExpiredCertPath := filepath.Join(tempBaseDir, "all_expired_cert.pem")
+	require.NoError(t, ioutil.WriteFile(allExpiredCertPath, allExpiredCertBundle, 0644))
+	// Now only use expired certs in the bundle, we should fail
+	err = ValidateRoot(certStore, newTestSignedRoot, "notary-signer", TrustPinConfig{CA: map[string]string{"notary-signer": allExpiredCertPath}, DisableTOFU: true})
+	require.Error(t, err)
+
+	// Add a CA cert for a that won't validate against the root leaf certificate
+	certStore.RemoveAll()
+	testPubKey3, err := cryptoService.Create("root", "notary-signer", data.ECDSAKey)
+	require.NoError(t, err)
+	testPrivKey3, _, err := memKeyStore.GetKey(testPubKey3.ID())
+	require.NoError(t, err)
+	validCert, err := cryptoservice.GenerateCertificate(testPrivKey3, "notary-signer", time.Now(), time.Now().AddDate(1, 0, 0))
+	require.NoError(t, err)
+	bundleWithWrongCert, err := trustmanager.CertChainToPEM([]*x509.Certificate{validCert})
+	require.NoError(t, err)
+	bundleWithWrongCertPath := filepath.Join(tempBaseDir, "bundle_with_expired_cert.pem")
+	require.NoError(t, ioutil.WriteFile(bundleWithWrongCertPath, bundleWithWrongCert, 0644))
+	err = ValidateRoot(certStore, newTestSignedRoot, "notary-signer", TrustPinConfig{CA: map[string]string{"notary-signer": bundleWithWrongCertPath}, DisableTOFU: true})
+	require.Error(t, err)
+}
+
```

Comment:
Here's a list of the new `certs_test.go` test cases (more or less in order) to make it easier to understand the coverage:

#### TOFUS
- Failure when tofus is disabled, and no other trust pinning is enabled
- Success when (inherited from other tests) tofus is enabled by default, and will succeed in bootstrapping trust when no other trust pinning is specified

#### Certs
- Success when the only provided valid cert ID is pinned for the specified GUN
- Success when only one of the provided cert IDs is valid for the specified GUN
- Success when none of the entries in `certs` match the GUN, and we fall through to TOFUS
- Failure when an invalid cert ID is pinned against the specified GUN
- Failure when an empty cert ID is pinned against the specified GUN
- Failure when the correct cert ID is pinned against the incorrect GUN (tofus explicitly disabled)
- Failure when a different key ID (targets key) is pinned against the specified GUN

#### CA
- Success when a valid CA bundle with a single valid root is pinned to the specified GUN
- Success when a valid CA bundle with one expired and one valid root is pinned to the specified GUN
- Failure when a CA bundle of only expired content is pinned to the specified GUN
- Failure when the path for the CA pinned against the specified GUN is invalid
- Failure when there are no pinned CAs for the GUN (tofus explicitly disabled)
- Failure when invalid PEM is pinned as CA for specified GUN
- Failure when a pinned CA for the GUN does not validate against the provided signed root